cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+  }
+
+  /**
+   * @brief Returns the AST to apply on bloom filters mmebership.
+   *
+   * @return AST operation expression
+   */
+  [[nodiscard]] std::reference_wrapper<ast::expression const> get_bloom_filter_expr() const
+  {
+    return _bloom_filter_expr.value().get();
+  }
+
+ private:
+  std::vector<cudf::size_type> _col_literals_offsets;
+};
+
+/**
+ * @brief Asynchronously reads bloom filters data to device.
+ *
+ * @param sources Dataset sources
+ * @param num_chunks Number of total column chunks to read
+ * @param bloom_filter_data Devicebuffers to hold bloom filter bitsets for each chunk
+ * @param bloom_filter_offsets Bloom filter offsets for all chunks
+ * @param bloom_filter_sizes Bloom filter sizes for all chunks
+ * @param chunk_source_map Association between each column chunk and its source
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return A future object for reading synchronization
+ */
+std::future<void> read_bloom_filter_data_async(
+  host_span<std::unique_ptr<datasource> const> sources,
+  size_t num_chunks,
+  cudf::host_span<rmm::device_buffer> bloom_filter_data,
+  cudf::host_span<std::optional<int64_t>> bloom_filter_offsets,
+  cudf::host_span<std::optional<int32_t>> bloom_filter_sizes,
+  std::vector<size_type> const& chunk_source_map,
+  rmm::cuda_stream_view stream)
+{
+  // Read tasks for bloom filter data
+  std::vector<std::future<size_t>> read_tasks;
+
+  // Read bloom filters for all column chunks
+  std::for_each(
+    thrust::make_counting_iterator<size_t>(0),
+    thrust::make_counting_iterator<size_t>(num_chunks),
+    [&](auto const chunk) {
+      // If bloom filter offset absent, fill in an empty buffer and skip ahead
+      if (not bloom_filter_offsets[chunk].has_value()) {
+        bloom_filter_data[chunk] = {};
+        return;
+      }
+      // Read bloom filter iff present
+      auto const bloom_filter_offset = bloom_filter_offsets[chunk].value();
+
+      // If Bloom filter size (header + bitset) is available, just read the entire thing.
+      // Else just read 256 bytes which will contain the entire header and may contain the
+      // entire bitset as well.
+      auto constexpr bloom_filter_size_guess = 256;
+      auto const initial_read_size =
+        static_cast<size_t>(bloom_filter_sizes[chunk].value_or(bloom_filter_size_guess));
+
+      // Read an initial buffer from source
+      auto& source = sources[chunk_source_map[chunk]];
+      auto buffer  = source->host_read(bloom_filter_offset, initial_read_size);
+
+      // Deserialize the Bloom filter header from the buffer.
+      BloomFilterHeader header;
+      CompactProtocolReader cp{buffer->data(), buffer->size()};
+      cp.read(&header);
+
+      // Get the hardcoded words_per_block value from `cuco::arrow_filter_policy` using a temporary
+      // `std::byte` key type.
+      auto constexpr words_per_block =
+        cuco::arrow_filter_policy<std::byte,
```

Comment:
Should we use `cuda::std::byte` to make the device code happier / more conformant?

Reply:
I think `std::byte` should be ok here as this is all host code. We only need the hardcoded `static constexpr words_per_block` variable from the policy here to verify that the number of bloom filter bitset bytes is a multiple of it.

Reply:
Changed to `cuda::std::byte`. `arrow_filter_policy<>::word_type` is also actually hardcoded to `uint32_t` 😆.

Reply:
Nope, not here. We actually instantiate and use it with the required `type` in `bloom_filter_caster::operator()` while instantiating a `bloom_filter_ref`